front/components/data_source/DocumentOrTableUploadOrEditModal.tsx

```diff
@@ -0,0 +1,486 @@
+  onClose: (save: boolean) => void;
+  owner: LightWorkspaceType;
+  plan: PlanType;
+}
+
+type TableOrDocument = {
+  name: string;
+  description: string;
+  file: File | null;
+  text: string;
+  tags: string[];
+  sourceUrl: string;
+};
+
+export function DocumentOrTableUploadOrEditModal({
+  dataSourceView,
+  contentNode,
+  isOpen,
+  onClose,
+  owner,
+  plan,
+}: DocumentOrTableUploadOrEditModalProps) {
+  const sendNotification = useContext(SendNotificationsContext);
+  const fileInputRef = useRef<HTMLInputElement>(null);
+
+  const [tableOrDoc, setTableOrDoc] = useState<TableOrDocument>({
+    name: "",
+    description: "",
+    file: null,
+    text: "",
+    tags: [],
+    sourceUrl: "",
+  });
+
+  const [uploading, setUploading] = useState(false);
+  const [isBigFile, setIsBigFile] = useState(false);
+  const [isLoading, setIsLoading] = useState(false);
+  const [developerOptionsVisible, setDeveloperOptionsVisible] = useState(false);
+
+  const isTable = contentNode?.type === "database";
+  const initialId = contentNode?.internalId;
+
+  const { table } = useTable({
+    workspaceId: owner.sId,
+    dataSourceName: dataSourceView.dataSource.name,
+    tableId: isTable ? initialId ?? null : null,
+  });
+
+  function resetTableOrDoc() {
+    setTableOrDoc({
+      name: "",
+      description: "",
+      file: null,
+      text: "",
+      tags: [],
+      sourceUrl: "",
+    });
+  }
+
+  useEffect(() => {
+    if (initialId) {
+      setIsLoading(true);
+      if (!isTable) {
+        setUploading(true);
+        fetch(
+          `/api/w/${owner.sId}/data_sources/${
+            dataSourceView.dataSource.name
+          }/documents/${encodeURIComponent(initialId)}`
+        )
+          .then(async (res) => {
+            if (res.ok) {
+              const document = await res.json();
+              setTableOrDoc((prev) => ({
+                ...prev,
+                name: initialId,
+                text: document.document.text,
+                tags: document.document.tags,
+                sourceUrl: document.document.source_url,
+              }));
+            }
+          })
+          .catch((e) => console.error(e))
+          .finally(() => setUploading(false));
+      } else if (table) {
+        setTableOrDoc((prev) => ({
+          ...prev,
+          name: table.name,
+          description: table.description,
+        }));
+      }
+      setIsLoading(false);
+    } else {
+      resetTableOrDoc();
+    }
+  }, [isTable, initialId, table, owner.sId, dataSourceView.dataSource.name]);
+
+  const total = 0; // TODO: Get the total number of documents
+
+  if (
+    !initialId &&
+    plan.limits.dataSources.documents.count !== -1 &&
+    total >= plan.limits.dataSources.documents.count
+  ) {
+    return (
+      <DocumentLimitPopup
+        isOpen={isOpen}
+        plan={plan}
+        onClose={() => onClose(false)}
+        owner={owner}
+      />
+    );
+  }
+
+  const handleUpload = async () => {
```

Comment:
It'd probably be clearer to create two functions here one for document, one for table so it removes a lot of conditional code.